python/lsst/daf/butler/_butler.py

```diff
@@ -600,8 +605,8 @@ def _findDatasetRef(self, datasetRefOrType: Union[DatasetRef, DatasetType, str],
 
         # Process dimension records that are using record information
         # rather than ids
-        newDataId: dict[Any, Any] = {}
-        byRecord: dict[Any, dict[str, Any]] = defaultdict(dict)
+        newDataId: Dict[Any, Any] = {}
```

Comment:
Oh, I think you mentioned something about proper types for data Id keys and values?

Reply:
This was a bit trickier than I expected, due to the messiness of `Mapping[Union[str, Dimension], ...]` vs. `Union[Mapping[str, ...], Mapping[Dimension, ...]]`.  Rather than fight that I just made these dicts use `str` keys and coerced the one place where they might have been given `Dimension` instances.